Enable index only scan on ao/aocs table.

haolinw opened the pull request:
Index scan has been disabled on ao/aocs table for that random
access performance on the ao/aocs table is extremely low.
While index only scan could save cycles from no access
of append-optimized varblocks, but only randomly reads
auxiliary Heap tables. Hence it is more performant than
index scan. Based on this theory, we enable index only scan
on append-optimized tables to serve satisfied queries that
no need to fetch tuples from physical append-optimized data
files. Check README for detail information.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

haolinw commented:
This is trying to refactor PR https://github.com/greenplum-db/gpdb/pull/10963

SmartKeyerror reviewed (commented):
We introduced another Bitmap to improve the efficiency of checking the visibility of a tuple, can we apply this to the normal scan? Such as SeqScan (in `aocs_getnext()`). 

If we can do that, can we get benefit from this?

haolinw commented on src/backend/executor/nodeIndexonlyscan.c:
```diff
@@ -90,11 +90,11 @@ IndexOnlyNext(IndexOnlyScanState *node)
-								   estate->es_snapshot,
-								   node->ioss_NumScanKeys,
-								   node->ioss_NumOrderByKeys);
+		scandesc = indexonly_beginscan(node->ss.ss_currentRelation,
```
No, we don't, missed that, thanks.

haolinw commented on src/backend/executor/nodeIndexonlyscan.c:
```diff
@@ -158,9 +158,8 @@ IndexOnlyNext(IndexOnlyScanState *node)
-		if (!VM_ALL_VISIBLE(scandesc->heapRelation,
-							ItemPointerGetBlockNumber(tid),
-							&node->ioss_VMBuffer))
+		if (!scandesc->xs_heapfetch->rel->rd_tableam->tid_visible(
```
Absolutely, will add table_tuple_tid_visible() wrapper.

haolinw commented on src/backend/optimizer/path/costsize.c:
```diff
@@ -690,6 +691,10 @@ cost_index(IndexPath *path, PlannerInfo *root, double loop_count,
 							  &spc_random_page_cost,
 							  &spc_seq_page_cost);
 
+	if (baserel_orig->amhandler == AO_ROW_TABLE_AM_HANDLER_OID ||
```
For AO tables, it should not be affected as pages_fetched will be always 0.

haolinw commented:
> We introduced another Bitmap to improve the efficiency of checking the visibility of a tuple, can we apply this to the normal scan? Such as SeqScan (in `aocs_getnext()`).
> 
> If we can do that, can we get benefit from this?

That's in common AppendOnlyVisimap_IsVisible() function, could be reached at need. Will check it for sure.

haolinw commented on src/include/cdb/cdbaocsam.h:
```diff
@@ -217,6 +217,8 @@ typedef struct AOCSFetchDescData
 	 * last sequence from gp_fastsequence.
 	 */
 	int64			lastSequence[AOTupleId_MultiplierSegmentFileNum];
+	int64 			lastRowNum[AOTupleId_MultiplierSegmentFileNum];
```
You mean firstRowNum[] and lastRowNum[] ? I think they should stand for the array of start and end row index for all segment files. Will add comments for them.

haolinw commented on src/backend/optimizer/path/indxpath.c:
```diff
@@ -809,11 +809,14 @@ get_index_paths(PlannerInfo *root, RelOptInfo *rel,
 		 * able to use index only scans on AO/AOCO relations. However it is
```
The GPDB_12_MERGE_FIXME could be removed as it could be addressed by this PR.

haolinw commented on src/include/access/tableam.h:
```diff
@@ -1069,6 +1069,14 @@ extern bool table_index_fetch_tuple_check(Relation rel,
 										  bool *all_dead);
 
+static inline bool
+tuple_index_fetch_tuple_visible(struct IndexFetchTableData *scan,
```
Could it be table_tuple_fetch_row_version() instead of introducing this new one ?

soumyadeep2007 commented:
Hi @haolinw , can you rebase your branch, squash and finalize the commits? I will take a look at this for sure starting Monday.

haolinw commented on src/backend/optimizer/path/costsize.c:
```diff
@@ -690,6 +691,10 @@ cost_index(IndexPath *path, PlannerInfo *root, double loop_count,
 
+	if (baserel_orig->relam == AO_ROW_TABLE_AM_OID||
+		baserel_orig->relam == AO_COLUMN_TABLE_AM_OID)
+		baserel->allvisfrac = 1.0;
```
Done.

soumyadeep2007 replied:
And also lets add as last line to the `If it's an index-only scan,` paragraph below:
`GPDB: For AO/CO tables, in an index-only scan, the base table never has to be scanned. So we ensure that allvisfrac=1.`

haolinw commented:
Thanks @soumyadeep2007 for following up this, I rebased it and will try to catchup your later comments.

ashwinstar commented:
> Another problem we encountered was the poor random access
performance of the visibility map. So we added a bitmap to
record the tuple id range of all the tuples visible.

Would be helpful to describe the problem and design of this more and possibly have separate commit for it as is pure performance optimization which is separate from rest of the work in this PR.

haolinw commented:
> > Another problem we encountered was the poor random access
> > performance of the visibility map. So we added a bitmap to
> > record the tuple id range of all the tuples visible.
> 
> Would be helpful to describe the problem and design of this more and possibly have separate commit for it as is pure performance optimization which is separate from rest of the work in this PR.

Yeah, that makes sense. Separated it.

kainwen commented on src/backend/access/appendonly/appendonly_visimap_entry.c:
```diff
@@ -459,6 +460,8 @@ AppendOnlyVisimapEntry_IsVisible(
 			   "(firstRowNum, rowNum) = (" INT64_FORMAT ", " INT64_FORMAT ")",
 			   visiMapEntry->firstRowNum, rowNum);
+		rangeNum = rowNum / APPENDONLY_VISIMAP_MAX_RANGE;
+		*allVisibleSet = bms_add_member(*allVisibleSet, rangeNum);
```
bitmapset might not be suitable for this case.

1. until now, it has overflow issue, since rangeNum is 64bit here, previous issue https://github.com/greenplum-db/gpdb/pull/12277
2. the comment at the head of `bitmapset.c` mentions the data structure is intended for max value is about a few hundred. Maybe feed large ints to it lead to perf issue? (Not very sure, need time to analyze the complexity)

haolinw replied:
One more https://github.com/greenplum-db/gpdb/pull/14529

soumyadeep2007 commented:
> > > Another problem we encountered was the poor random access
> > > performance of the visibility map. So we added a bitmap to
> > > record the tuple id range of all the tuples visible.
> > 
> > 
> > Would be helpful to describe the problem and design of this more and possibly have separate commit for it as is pure performance optimization which is separate from rest of the work in this PR.
> 
> Yeah, that makes sense. Separated it.

Can you squash the changes or remove the commit "Use a fixed-length FileSegInfo* Array in AppendOnlyFetchDesc" ?
I can't quite follow what is going on.

haolinw commented:
> > > > Another problem we encountered was the poor random access
> > > > performance of the visibility map. So we added a bitmap to
> > > > record the tuple id range of all the tuples visible.
> > > 
> > > 
> > > Would be helpful to describe the problem and design of this more and possibly have separate commit for it as is pure performance optimization which is separate from rest of the work in this PR.
> > 
> > 
> > Yeah, that makes sense. Separated it.
> 
> Can you squash the changes or remove the commit "Use a fixed-length FileSegInfo* Array in AppendOnlyFetchDesc" ? I can't quite follow what is going on.

Done. I removed "Use a fixed-length FileSegInfo* Array" in my refactor version, instead, introducing two int arrays for the segno<->idx mapping purpose, refer to src/include/access/appendonlytid.h

haolinw commented:
Need to add a case to test a question from @soumyadeep2007: if the first insert operation was aborted after creating table, currently no data in visimap but has dead tuples in table, indexonlyscan still works?

soumyadeep2007 commented:
> Need to add a case to test a question from @soumyadeep2007: if the first insert operation was aborted after creating table, currently no data in visimap but has dead tuples in table, indexonlyscan still works?

I added a sq me commit with a targeted uao test. This scenario passes and other visibility scenarios added passes for both AO and CO.

ashwinstar commented:
> The index scan has disabled on ao/aocs table for that
random access performance on the ao/aocs table is extremely
low. But index only scan does not access data files. Its
performance should be as good as the heap table. So index
only scan should be enable on ao/aocs table.

I would like to reflect on this aspect, just access blocks on disk being random isn't the reason for disabling index scans. Even visimap lookups in random order could yield to poor performance. We have seen this for vacuum task where index cleanup was performed based on visibility map and performance was not good at all. Thanks to PR https://github.com/greenplum-db/gpdb/pull/13255 which removed dependency on visimap to make that decision during vacuum's. Hence, need to quantify this aspect to possible to factor in when index-only scans should be picked for append-optimized tables.

ashwinstar commented:
Would be also helpful to understand the costing or factors based on which index-only scan will be picked for append-optimized tables (decision used for heap may not be relevant to make plan choice for AO) after this PR.

soumyadeep2007 reviewed (commented):
Discussed with @ashwinstar and @haolinw 

(1) Please defer the segno optimization to a completely separate PR and remove from this PR entirely. Seg files don't have to be consulted for index only scans.

(2) Tests should be of 2 different categories: Testing correctness and testing planning decisions.
Comments:
Run the newly added test w/ `./pg_isolation2_regress --init-file=../regress/init_file --init-file=init_file_isolation2 --load-extension=gp_inject_fault uao/index_only_scan_row uao/index_only_scan_column`

(3) High-level workflow should be:

Setup sysscan of block directory once for entire Index only scan
Setup visimapscan once for entire Index only scan

Use sysscan of block directory to check if a given tid is visible. 
(Try to see if AppendOnlyBlockDirectoryEntry, set_directoryentry_range can be used for better perfomance - consider later) 

	If not visible, return false and index only scan should skip this tuple in result set.

	If visible, then we need to check visimap and then return accordingly.

End sysscan of block directory at Index only scan end.
End visimapscan at Index only scan end.

Comments:
* We talked about borrowing some of the code or creating similar function to `blkdir_entry_exists`, which was introduced for uniqueness checks. We can't directly use it since it assumes a dirty snapshot and starts a scan -> scans -> stops the scan on every call.
* AppendOnlyBlockDirectory_GetEntry should not be used as it can sometimes return true even though a block directory doesn't exist (See how we have to call `scanToFetchTuple` additionally in `appendonly_fetch` to determine visibility)

(4) Anti-goals:

Scanning the actual AO tuple from the AO relation. This is not necessary as
visibility information is not stored on AO tuples. **This is currently happening due to the code structure .. `index_fetch_heap` is being called for aborted and deleted tuples**

ashwinstar commented:
> Please defer the segno optimization to a completely separate PR and remove from this PR entirely. Seg files don't have to be consulted for index only scans.

This is very important aspect to pay close attention too. In past (before 258ec96), pg_aoseg/pg_aocsseg stored EOF (or checking physical file for tuple presence) was the only source to perform transaction visibility checks for append-optimized tables (blockdirectory was only used for optimization). 

With the change made by commit 258ec96 to align block directory reflect the reality of block information on-disk, now the design is - this is something we should very clearly mention in README if not already
- for non-indexed append-optimized tables EOF acts as source (along with visimap)
- for indexed append-optimized tables blockdirectory acts as source (along with visimap)

Commit 258ec96 changed the format version to `AORelationVersion_PG12`, so only tables with this format version should enable index-only scan using the above logic. So, tables in-place upgraded from GPDB6->GPDB7 with old block directory format should not pick index-only scans.

soumyadeep2007 commented:
> Also, lets make sure to update [README](https://github.com/greenplum-db/gpdb/blob/main/src/backend/access/appendonly/README.md) with all the design for this feature

Yes, let's create a section in the readme titled "Index Only Scan", similar to the one created for unique indexes.

soumyadeep2007 commented:
> Commit [258ec96](https://github.com/greenplum-db/gpdb/commit/258ec966b26929430fc5dc9f6e6fe09854644302) changed the format version to `AORelationVersion_PG12`, so only tables with this format version should enable index-only scan using the above logic. So, tables in-place upgraded from GPDB6->GPDB7 with old block directory format should not pick index-only scans.

You might be able to reuse `ValidateRelationVersionForUniqueIndex` for this. Please rename the function appropriately.

haolinw commented:
> > Commit [258ec96](https://github.com/greenplum-db/gpdb/commit/258ec966b26929430fc5dc9f6e6fe09854644302) changed the format version to `AORelationVersion_PG12`, so only tables with this format version should enable index-only scan using the above logic. So, tables in-place upgraded from GPDB6->GPDB7 with old block directory format should not pick index-only scans.
> 
> You might be able to reuse `ValidateRelationVersionForUniqueIndex` for this. Please rename the function appropriately.

Got it. Thanks Deep.

haolinw commented:
> (1) Please defer the segno optimization to a completely separate PR and remove from this PR entirely. Seg files don't have to be consulted for index only scans.

Separate segno lookup optimization to a new PR https://github.com/greenplum-db/gpdb/pull/14541.

haolinw commented on src/include/cdb/cdbappendonlyam.h:
```diff
@@ -321,6 +321,16 @@ typedef struct AppendOnlyFetchDescData
+	 */
+	int64 			lastRowNum[AOTupleId_MultiplierSegmentFileNum];
+	int64 			firstRowNum[AOTupleId_MultiplierSegmentFileNum];
+
```
Removed them.

haolinw commented:
> Commit [258ec96](https://github.com/greenplum-db/gpdb/commit/258ec966b26929430fc5dc9f6e6fe09854644302) changed the format version to `AORelationVersion_PG12`, so only tables with this format version should enable index-only scan using the above logic. So, tables in-place upgraded from GPDB6->GPDB7 with old block directory format should not pick index-only scans.

Have a question on this:
Seems there is no straightforward place to inject code to determine whether select indexonlyscan or not based on the format version of AO blkdir during planning phase (which needs I/O to access pg_aoseg). Here lists possible ways I can see to move forward to determine the final implementation:
1. Inject code to estimate_rel_size() which seems the only place where could call GetFileSegInfo() to get the formatversion in the planer layer, then the later paths selection might be able to determine indexonlyscan based on that. This seems to be the best option if it can be implemented properly.
2. If above option is not feasible, the indexonlyscan might be selected without knowing the underlying blkdir format, then error out if indexonlyscan in a pre-AORelationVersion_PG12 format, just like unique index does (`ValidateRelationVersionForUniqueIndex`). This is the simplest option, but might introduce failure to upgrade hence documentation and workaround should be considered clearly. 
3. More complicated, not error out in above case, instead, following the original logic to do physical scanning (call `scanToFetchTuple`) after visimap check to get the final visibility of the tuple. This is similar like Heap behavior and seems more friendly than error. I saw we have already ruled out this option from previous comments, right ? List here to confirm.

I would try to implement the 1st option, if not success, then fall back to 2nd. Please let me know your advice or concerns. @ashwinstar @soumyadeep2007

haolinw commented on src/backend/access/appendonly/appendonlyam.c:
```diff
@@ -2415,6 +2418,49 @@ appendonly_fetch(AppendOnlyFetchDesc aoFetchDesc,
+	GET_SEGINFO(aoFetchDesc->segmentFileInfo, segmentFileNum, segInfo);
+
+	if (curBlockDirectoryEntry->range.afterFileOffset > segInfo->eof)
+		return false;
```
Done.

haolinw commented:
> 1. Inject code to estimate_rel_size() which seems the only place where could(or already) call GetFileSegInfo() to get the formatversion in the planer layer, then the later paths selection might be able to determine indexonlyscan based on that. This seems to be the best option if it can be implemented properly.

Failed to implement option 1 as no chance to read segment pg_aoseg info during query planning in coordinator, and don't think it is a good idea to introduce segment I/O here to determine whether picking indexonlyscan path or not. If I understand correctly, seems this option is not feasible.

haolinw commented:
> > Also, lets make sure to update [README](https://github.com/greenplum-db/gpdb/blob/main/src/backend/access/appendonly/README.md) with all the design for this feature
> 
> Yes, let's create a section in the readme titled "Index Only Scan", similar to the one created for unique indexes.

Will do that once finalize all designing aspects.

soumyadeep2007 reviewed (commented):
We need the following refactorings **which we should tackle as a separate PR first**:

1. Rename `AppendOnlyBlockDirectory_CoversTuple() ->
	AppendOnlyBlockDirectory_Unique_Check()`
Please do this in a separate commit.

2. We need a way to store the blkdir sysscan desc in the
`AppendOnlyBlockDirectory` struct (`AppendOnlyBlockDirectory->scan`). We should
refactor `AppendOnlyBlockDirectory_GetEntry()` to not set up the sysscan desc
multiple times, and use this struct member. We can initialize it in
`AppendOnlyBlockDirectory_Init_forSearch()`. This struct member will be used
later on in our IndexOnlyScan code.
Please do this in a separate commit.

3. Rename `table_index_fetch_tuple_exists, index_fetch_tuple_exists` to
`table_index_unique_check and index_unique_check` respectively.

Please ensure that the README and comments are updated with the changes from the refactoring.

These refactorings will help a lot for our implementation I feel as described below.

Then with that PR in, we should introduce the following:

* Table AM API:
`bool table_index_fetch_tuple_visible(Relation rel, ItemPointer tid, Snapshot snapshot)`

* We shouldn't need to call `*_fetch_init()` as that does too much initialization.
Those functions are meant to perform tuple fetches from the actual table, and
are used for regular index scans.
    * Now in terms of initialization, we ONLY need to initialize the 
i) block directory ii) visimap
to perform the tuple visibility check, much like unique index checks.
    * So, we should write a *_fetch_index_only_init() function which initializes just
the blkdir and visimap.
    * To initialize the block directory, we can simply write a variant of `AppendOnlyBlockDirectory_Init_forUniqueChecks()`:
         ```
          void
          AppendOnlyBlockDirectory_Init_forIndexOnlyScan(AppendOnlyBlockDirectory *blockDirectory, Relation aoRel, int numColumnGroups, Snapshot snapshot)
          {
	          // Will look the SAME as AppendOnlyBlockDirectory_Init_forUniqueChecks()
	          // with the following changes:
	          
	          // 1. Remove existing snapshot asserts. You can add an assert stating that
	          //    the snapshot is an MVCC snapshot. See comment in nodeIndexOnlyScan.c:
	          //    "Only MVCC snapshots are supported here..."
	          //
	          // 2. Slight adjustment to elog(Debug_appendonly_print_blockdirectory,..)
	          //
	          // 3. We will be setting the snapshot here.
	          //    blockDirectory->appendOnlyMetaDataSnapshot = snapshot;
	          //
	          // 4. Set up AppendOnlyBlockDirectory->scan with sysscan init functions like
	          //    we do in blkdir_entry_exists(), but with the snapshot supplied.
          
          }
         ```
    * Similarly, we should have `AppendOnlyBlockDirectory_End_forIndexOnlyScan()`.
    * For the visimap, we should do the same thing - write variants of `AppendOnlyVisimap_Init_forUniqueCheck()` and `AppendOnlyVisimap_Finish_forUniquenessChecks()`
    * Inside the AM implementation, to perform a visibility check we should NOT call
AppendOnlyBlockDirectory_GetEntry(), as it can fetch the underlying tuple from
the relfile! Instead, we can simply call:
`AppendOnlyBlockDirectory_IndexOnly_Check(AppendOnlyBlockDirectory *blockDirectory, AOTupleId *aoTupleId)`
which will internally call:
a) For AO:       `blkdir_entry_exists(blockDirectory, aoTupleId, 0)`
b) For AOCO:  `blkdir_entry_exists(blockDirectory, aoTupleId, <any one index attribute>)`
    * We will need to refactor `blkdir_exists` so that: 
         * For uniqueness checks it always initializes/destroys a fresh sysscan descriptor.
         * For index only scan it reuses `blockDirectory->scan`.

haolinw commented on src/include/cdb/cdbaocsam.h:
```diff
@@ -231,6 +231,9 @@ typedef struct AOCSFetchDescData
 
 	Oid segrelid;
+
+	/* Assuming only one formatversion for all segfiles. */
+	int16 formatversion;
```
Sorry for late response. Still haven't got a chance to look into detail, add it to my plan with next high priority task and will schedule a big time slot focusing on it.

haolinw replied:
Hi @soumyadeep2007 ,  I'm trying to fix this incorrect assumption by using `minformatversion` which only stores the minimum formatversion for all segfiles instead, and keep `need_fetch_tuple` assignment and checking logic with additional `WARNING` message for reporting falling back to indexscan from indexonlyscan and hint user how to fix.

Would it work as a temporary fix for this comment before finalizing the tricky part solution in optimizer ?

soumyadeep2007 replied:
> Hi @soumyadeep2007 , I'm trying to [fix](https://github.com/greenplum-db/gpdb/pull/13766/commits/e8f0a22a526866acceadf7a02082a48f3d8f26f5) this incorrect assumption by using `minformatversion` which only stores the minimum formatversion for all segfiles instead, and keep `need_fetch_tuple` assignment and checking logic with additional `WARNING` message for reporting falling back to indexscan from indexonlyscan and hint user how to fix.
> 
> Would it work as a temporary fix for this comment before finalizing the tricky part solution in optimizer (maybe in another PR)?


Chatted with @ashwinstar today. Our take on this that we should open a separate PR to record a version field in pg_appendonly first (Option 2 above) and then use that version in this PR to ban plans. We don't the like the idea of introducing this as we don't want to fall back to IndexScan which is very inefficient for AO/CO tables.

haolinw replied:
Got it, thanks, will try to incorporate with this.

haolinw replied:
By the way, I'm also thinking about blkdir related comments, I feel the following one may duplicate with fast-analyze f123f5a7aa8335aedac59da93f817093a93c13fd. I'm also asking for that PR review, would you mind if also take a look about that and consider both blkdir changes together then maybe we can generate a new unified interface to apply to both PRs ? Since I think the existing AppendOnlyBlockDirectory_GetEntry() was used on critical path stably for long time hence any change to that needs to take special caution, while introducing new interface for new scenario may limit the scope of impact.

> We need a way to store the blkdir sysscan desc in the
AppendOnlyBlockDirectory struct (AppendOnlyBlockDirectory->scan). We should
refactor AppendOnlyBlockDirectory_GetEntry() to not set up the sysscan desc
multiple times, and use this struct member. We can initialize it in
AppendOnlyBlockDirectory_Init_forSearch(). This struct member will be used
later on in our IndexOnlyScan code.
Please do this in a separate commit.

soumyadeep2007 replied:
> By the way, I'm also thinking about blkdir related comments, I feel the following one may duplicate with fast-analyze https://github.com/greenplum-db/gpdb/commit/f123f5a7aa8335aedac59da93f817093a93c13fd. I'm also asking for that PR review, would you mind if also take a look about that and consider both blkdir changes together then maybe we can generate a new unified interface to apply to both PRs ?

I haven't looked at it yet. I will take a look and think about it. But we might be able to have a unified interface - just like we can't unify the API for unique checks with this PR. Refer my [comment](https://github.com/greenplum-db/gpdb/pull/13766#pullrequestreview-1266599546) above. **We should focus on this PR first and then start the fast analyze PR, I strongly feel.** If there is any unification that needs to be done, we can also do it as a follow-up. However, I will do a quick check and get back to you.

> Since I think the existing AppendOnlyBlockDirectory_GetEntry() was used on critical path stably for long time hence any change to that needs to take special caution, while introducing new interface for new scenario may limit the scope of impact. (Understand optimization on AppendOnlyBlockDirectory_GetEntry() would get performance benefit and may be proper to be done in later independent work after this ? Actually we did separate [other performance related changes](https://github.com/greenplum-db/gpdb/pull/14541) from this work).

Hmmm, I would agree with you if this was any other case. But think of it like this, if we scan the tuple from the relfile as a result of `AppendOnlyBlockDirectory_GetEntry()`, then the feature is no longer index-only-scan (as we are scanning the table) and it defeats the whole purpose of this PR. Right?

haolinw replied:
Sorry, I don't get that. I didn't see `AppendOnlyBlockDirectory_GetEntry` introducing extra AO table scanning but just blkdir (aux heap table) indexscan which works similar like `AppendOnlyVisimapStore_Find`.

haolinw replied:
I see the only reason needing to fetch tuple from AO relfile is because the minimum AO format version is less than AORelationVersion_PG12. Did I misunderstanding something ?

soumyadeep2007 replied:
I'm sorry I misremembered. This is the reason we can't use GetEntry, because it can return true in some cases even when a minipage entry could not be found:

```
if (entry_no == -1)
{
	/*
		* Since the last few blocks may not be logged in the block
		* directory, we always use the last entry.
		*
		* FIXME: If we didn't find a suitable entry, why even use the last
		* entry? Currently, as it stands we would most likely return
		* true from this function. This will lead to us having to do a
		* fetch of the tuple from the physical file in the layer above (see
		* scanToFetchTuple()), where we would ultimately find the tuple
		* missing. Would it be correct to set the directory entry here to
		* be the last one (for caching purposes) and return false, in order
		* to avoid this physical file read?
		*/
	entry_no = minipageInfo->numMinipageEntries - 1;
}
return set_directoryentry_range(blockDirectory,
								columnGroupNo,
								entry_no,
								directoryEntry);
```

@ashwinstar and I tried removing this logic during the unique index work but we couldn't.

haolinw replied:
It's interesting, I read that before but didn't get in which case could let last few blocks be not recorded in blkdir, can you show me an example ? I understood crash insert may leave larger block range than physical file.

soumyadeep2007 replied:
I unfortunately don't remember. I think you can repro it with one of the tests in isolation2 or regress. That's what I remember doing too.

ashwinstar replied:
I don't remember the details either. Though, I don' think its case of last few blocks are not recorded or missed in blkdir. If I remember correctly, its most have to do with how this interface is used by callers to return last block directory if none matching the block is found and something on those lines. As Deep said, existing test did provide clue why the logic is needed.

haolinw replied:
OK, summarize what I'm understanding to sync:
- if AO segments are all on formatversion >= AORelationVersion_PG12, blkdir entries could cover all ranges of the physical segfile, no missing `last few blocks`, no gaps in blkdir entry, so we can trust `GetEntry` result in this case;
- if AO segments contain a segfile of which formatversion < AORelationVersion_PG12 due to upgrade, blkdir entry could still have gap, or missing `last few blocks` because of missing entries in blkdir, which may result to a case that `GetEntry` returns true (with last entry) but actually the requesting tuple is not existent, so we can't count on `GetEntry` result in this case.
- AORelationVersion_PG12 should be able to provide a consistent view between aoseg and blkdir catalogs which means: 
    - 1) if formatversion is mixed (i.e., seg1 < AORelationVersion_PG12, seg2 >= AORelationVersion_PG12) in aoseg, blkdir could miss entries (i.e. `last few blocks`) hence `GetEntry` shouldn't be trusted; 
    - 2) if all segfiles formatversion is >= AORelationVersion_PG12 in aoseg, no entry missing in blkdir hence `GetEntry` can be trusted.
    - 3) So, if a scan is initialized with all segfiles's formatversion checking via aoseg, the judgment result should be sufficient to determine trusting `GetEntry` or not for blkdirscan.

Is above correct ?

soumyadeep2007 replied:
Even with everything having formatversion >= PG12, GetEntry can lie. Running existing tests should help you find when GetEntry will lie.

haolinw replied:
Did you mean the case `firstRowNum < lastRowNum < rowNum` [here](https://github.com/greenplum-db/gpdb/commit/258ec966b26929430fc5dc9f6e6fe09854644302#diff-01ea7ae3ed9be5a98803959b6de841504d0a5403cafd8f144664ab5fb4964badR1946) ? 
If so, I should have corrected my above words more accurately, but it might not be what I was asking.

When all formatversion >=PG12, if GetEntry lies, the rowNum should be not in the range anymore, so I have similar [code](https://github.com/greenplum-db/gpdb/pull/13766/files#diff-01ea7ae3ed9be5a98803959b6de841504d0a5403cafd8f144664ab5fb4964badR2458) to prevent from this fault, it will treat the tuple as invisible, but not fall back to fetch tuple, so should be no impact to pure indexonlyscan. Otherwise, it could be conflict with this [words](https://github.com/greenplum-db/gpdb/commit/258ec966b26929430fc5dc9f6e6fe09854644302#diff-01ea7ae3ed9be5a98803959b6de841504d0a5403cafd8f144664ab5fb4964badR1929). 

I did spent almost an hour to look for the case but couldn't find the exact one what I expected even though I speculate it should be in ao_blkdir.sql.  Will continue to inspect that file but if my above conclusion is correct I should be fine to go with the conclusion ?

soumyadeep2007 replied:
>Did you mean the case firstRowNum < lastRowNum < rowNum [here](https://github.com/greenplum-db/gpdb/commit/258ec966b26929430fc5dc9f6e6fe09854644302#diff-01ea7ae3ed9be5a98803959b6de841504d0a5403cafd8f144664ab5fb4964badR1946) ?

Yes that case is a possibility. **But there could be more.** Broadly speaking, it is one of the cases where we hit the branch:
```c
/*
 * Since the last few blocks may not be logged in the block
 * directory, we always use the last entry.
 *
```
**and then return `true` from the `set_directoryentry_range()` call immediately below.** Then afterwards, `scanToFetchTuple` will scan the table and then `fetchNextBlock->AppendOnlyStorageRead_ReadNextBlock` could return false or the case you mentioned above.

> I did spent almost an hour to look for the case but couldn't find the exact one what I expected even though I speculate it should be in ao_blkdir.sql. Will continue to inspect that file but if my above conclusion is correct I should be fine to go with that ?

I think it is in some of the older files. `ao_blkdir.sql` was what we added recently. Try running all the uao tests in regress and isolation2 and add a log message to the branch:
```c
/*
 * Since the last few blocks may not be logged in the block
 * directory, we always use the last entry.
 *
```

> When all formatversion >=PG12, if GetEntry lies, the rowNum should not be in the range anymore, so I have similar [code](https://github.com/greenplum-db/gpdb/pull/13766/files#diff-01ea7ae3ed9be5a98803959b6de841504d0a5403cafd8f144664ab5fb4964badR2458) to prevent from this fault, it will treat the tuple as invisible, but not fall back to fetch tuple, so should be no impact to pure indexonlyscan.

I am not sure how the code (comment) you mentioned is relevant here. Can you please explain?

haolinw replied:
Under all formatversion >= PG12 context, I read the blkdir entry hole filling commit c93904f73e44ee96031d3e0681bc8de9abc5c1d7 provided an invariant `fetching physical
rows in the continuous range of a block directory entry's first and last row numbers will always be successful.` I thought even if GetEntry lies, double check range by RangeHasRow could identify this fault as GetEntry filled the last entry which didn't cover the requesting rowNum. And I thought `firstRowNum < lastRowNum < rowNum` is the only condition to hit this case when the requesting tuple is invisible as insert transaction was aborted. So I have that code to double check range after GetEntry to intend to prevent from being fault as a result of GetEntry. So the question is, can I say that RangeHasRow could resolve that and the requesting tuple should be treated as invisible in this situation ?

haolinw replied:
> I think it is in some of the older files. ao_blkdir.sql was what we added recently. Try running all the uao tests in regress and isolation2 and add a log message to the branch:

Thanks for the info, will try it out later.

haolinw replied:
> I am not sure how the code (comment) you mentioned is relevant here. Can you please explain?

Ah, seems the link wasn't pointing the right location, pasted relevant code below. Intended to prevent from GetEntry fault by double checking the range of the requesting tuple. The assertion is, if GetEntry returns true, the requesting tuple should be fall into the entry range [firstRowNum, lastRowNum], otherwise, the tuple should be rendered as invisible, correct ? Could it be possible that GetEntry returned true but not in the returned-entry range but the tuple had to be rendered as visible (subsequent `scanToFetchTuple` returned true) ?

```c
static inline bool
appendonly_tid_in_blkdir_entry(AppendOnlyFetchDesc aoFetchDesc,
				   			   AOTupleId *aoTupleId)
{
	int			segmentFileNum = AOTupleIdGet_segmentFileNum(aoTupleId);
	int64		rowNum = AOTupleIdGet_rowNum(aoTupleId);
	AppendOnlyBlockDirectoryEntry *entry = &aoFetchDesc->currentBlock.blockDirectoryEntry;

	return (segmentFileNum == aoFetchDesc->blockDirectory.currentSegmentFileNum &&
			AppendOnlyBlockDirectoryEntry_RangeHasRow(entry, rowNum));
}

bool
appendonly_tuple_visible(AppendOnlyFetchDesc aoFetchDesc,
						 AOTupleId *aoTupleId)
{
	bool got_entry = false;
	while (!appendonly_tid_in_blkdir_entry(aoFetchDesc, aoTupleId))
	{
		if (got_entry)
			return false;

		if(!AppendOnlyBlockDirectory_GetEntry(&aoFetchDesc->blockDirectory,
											  aoTupleId,
											  0,
											  &aoFetchDesc->currentBlock.blockDirectoryEntry))
			return false;

		got_entry = true;

		/*
		 * Go back to check aoTupleId if in new obtained blkdir entry.
		 * We have to do this because aoTupleId.rowRum might be greater
		 * than blkdir last entry.lastRowNum (check the comments at the
		 * end of AppendOnlyBlockDirectory_GetEntry()). If that's the
		 * case, currently we think the tuple is invisible.
		 */
	}

	if (aoFetchDesc->snapshot != SnapshotAny &&
		!AppendOnlyVisimap_IsVisible(&aoFetchDesc->visibilityMap, aoTupleId))
		return false;

	return true;
}
```

soumyadeep2007 replied:
> Could it be possible that GetEntry returned true but not in the returned-entry range but the tuple had to be rendered as visible (subsequent scanToFetchTuple returned true) ?

I am not sure. The existing code (all of the existing scanToFetchTuple and AppendOnlyBlockDirectory_GetEntry etc) is so indirect and poorly documented, its hard to say.

That is one of the reasons we opted to not touch/reuse these functions in our unique index work and let these be very specific to the scan code.

It's a bad excuse - I wish I had a better answer for you :)

haolinw replied:
Understand. Thank you all the same. Will continue to incorporate with previous comments to make things clear.

haolinw commented:
@soumyadeep2007 Let's move forward :-). I think the pg_appendonly part should be ready to merge. Next, I'm going to work on block directory refector. Some questions come up after review your previous comments below:
>  We need a way to store the blkdir sysscan desc in the
AppendOnlyBlockDirectory struct (AppendOnlyBlockDirectory->scan). We should
refactor AppendOnlyBlockDirectory_GetEntry() to not set up the sysscan desc
multiple times, and use this struct member. We can initialize it in
AppendOnlyBlockDirectory_Init_forSearch(). This struct member will be used
later on in our IndexOnlyScan code.

Don't quite understand why we need to cache the sysscan for indexscan on blkdir as the scan key is not always same per call which is not like seqscan (which I introdced and cache that scan in fast-analyze).

> Table AM API:
bool table_index_fetch_tuple_visible(Relation rel, ItemPointer tid, Snapshot snapshot)
We shouldn't need to call *_fetch_init() as that does too much initialization.
Those functions are meant to perform tuple fetches from the actual table, and
are used for regular index scans.

If we don't "fetch" tuple, is table_index_tuple_visible() a better name ?

soumyadeep2007 commented:
> Don't quite understand why we need to cache the sysscan for indexscan on blkdir as the scan key is not always same per call which is not like seqscan (which I introdced and cache that scan in fast-analyze).

You are right its not the same. I don't know how I overlooked that.

> If it is not necessary, I would not touch GetEntry and try to refactor AO indexonlyscan by using blkdir_entry_exists. Sounds OK ?

Sounds good!

> Also, if there's no work on GetEntry, is that OK to put the other two renaming comments(item 1&3 https://github.com/greenplum-db/gpdb/pull/13766#pullrequestreview-1266599546) in a separate commit in this PR (not a separate PR) ?

If you prefer, no problem!

> If we don't "fetch" tuple, is table_index_tuple_visible() a better name ?

Yes, but if you see `IndexOnlyNext()`, we call `index_beginscan()` which calls `table_index_fetch_begin()` and so on. So I wanted to add `table_index_fetch_tuple_visible()` to that family by keeping the `table_index_fetch` prefix. We can make things clear with comments.

haolinw commented:
> Yes, but if you see IndexOnlyNext(), we call index_beginscan() which calls table_index_fetch_begin() and so on. So I wanted to add table_index_fetch_tuple_visible() to that family by keeping the table_index_fetch prefix. We can make things clear with comments.

I see. That works for me for using `table_index_fetch_tuple_visible`.

haolinw commented:
I think I have addressed most of the previous comments, mainly including:
- ban plan based on pg_appendonly.version
- refactor indexonlyscan tableam based on `blkdir_entry_exists`

Please continue to review, thanks.  cc @soumyadeep2007 @ashwinstar.

soumyadeep2007 commented on src/test/regress/expected/bfv_index_optimizer.out:
```diff
@@ -485,41 +485,41 @@ set seq_page_cost=10000000;
+                                         QUERY PLAN                                          
+---------------------------------------------------------------------------------------------
+ Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..437.00 rows=1 width=16)
+   ->  Nested Loop  (cost=0.00..437.00 rows=1 width=16)
          Join Filter: true
-         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..431.00 rows=20 width=8)
-               ->  Seq Scan on nestloop_x  (cost=0.00..431.00 rows=7 width=8)
-         ->  Index Scan using nestloop_y_idx on nestloop_y  (cost=0.00..120.01 rows=1 width=8)
+         ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..431.00 rows=1 width=8)
+               ->  Seq Scan on nestloop_x  (cost=0.00..431.00 rows=1 width=8)
+         ->  Index Scan using nestloop_y_idx on nestloop_y  (cost=0.00..6.00 rows=1 width=8)
```
Why are these costs changing? Don't quite follow as we didn't change the cost model for regular index scans.

haolinw replied:
Same as above, actually no changes on costs.

soumyadeep2007 commented on src/test/regress/expected/bfv_index_optimizer.out:
```diff
@@ -529,8 +529,8 @@ select * from nestloop_x as x, nestloop_y as y where y.j > x.i + x.j + 2;
  1 | 1 | 6 | 6
- 1 | 1 | 7 | 7
  2 | 2 | 7 | 7
+ 1 | 1 | 7 | 7
```
Same thing, can be avoided.

haolinw replied:
Done.

soumyadeep2007 commented on src/test/regress/expected/bfv_index_optimizer.out:
```diff
@@ -1002,3 +1005,79 @@ DROP INDEX hash_idx3;
+CREATE INDEX bfv_index_only_ao_a_b on bfv_index_only_ao(a) include (b);
+insert into bfv_index_only_ao select i,i from generate_series(1, 10000) i;
+explain select count(*) from bfv_index_only_ao where a < 100;
+                                     QUERY PLAN                                     
+------------------------------------------------------------------------------------
+ Aggregate  (cost=0.00..431.00 rows=1 width=8)
+   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=1)
+         ->  Seq Scan on bfv_index_only_ao  (cost=0.00..431.00 rows=1 width=1)
+               Filter: (a < 100)
+ Optimizer: Pivotal Optimizer (GPORCA)
+(5 rows)
```
All the plans here are seqscan when we want IOscan.

soumyadeep2007 reviewed (commented):
Changes are looking good! Some more feedback added. Not major but important changes, I feel.

Few things:
* I don't think our cost model is comprehensive enough as we don't take into account cost of scanning aux tables. We should leave that as a FIXME in `cost_index` and leave that for another PR.
* There are some weird diffs unrelated to index only scan in the bfv answer files. I can't quite grok these.
* Please also apply this patch for more commentary around the block directory changes.
[blkdir_comments.patch.txt](https://github.com/greenplum-db/gpdb/files/10926735/blkdir_comments.patch.txt)

haolinw commented on src/backend/executor/nodeIndexonlyscan.c:
```diff
@@ -126,12 +126,14 @@ IndexOnlyNext(IndexOnlyScanState *node)
 				continue;
-		} /* CAUTION: else if branch is under the comments. */
+			else
+				; /* visible, jump to filling tuple slot logic */
```
Done.

haolinw commented on src/backend/executor/nodeIndexonlyscan.c:
```diff
@@ -126,12 +126,14 @@ IndexOnlyNext(IndexOnlyScanState *node)
+			else
+				; /* visible, jump to filling tuple slot logic */
+
+		} /* CAUTION: else branch is under the comments. */
```
Done.

haolinw commented on src/backend/optimizer/util/plancat.c:
```diff
@@ -190,6 +190,16 @@ get_relation_info(PlannerInfo *root, Oid relationObjectId, bool inhparent,
 
+		/*
+		 * GPDB needs to get AO relation version from pg_appendonly catalog to
+		 * determin whether enabling IndexOnlyScan on AO or not.
```
Done.

haolinw commented on src/include/cdb/cdbaocsam.h:
```diff
@@ -283,6 +289,8 @@ typedef struct IndexFetchAOCOData
 
 	AOCSFetchDesc       aocofetch;
 
+	AOCSIndexOnlyDesc	indexonlydesc;
```
Done.

haolinw commented on src/backend/access/aocs/aocsam.c:
```diff
@@ -1654,7 +1594,51 @@ aocs_fetch_finish(AOCSFetchDesc aocsFetchDesc)
+	return indexonlydesc;										
+}
+
+bool
+aocs_index_only_check(AOCSIndexOnlyDesc indexonlydesc, AOTupleId *aotid, Snapshot snapshot)
+{
+	if (!AppendOnlyBlockDirectory_CoversTuple(indexonlydesc->blockDirectory, aotid))
+		return false;
+
+	if (snapshot != SnapshotAny && !AppendOnlyVisimap_IsVisible(indexonlydesc->visimap, aotid))
+		return false;
+	
+	return true;
+}
```
I thought If gp_select_invisible is set, it might return false on a removed tuple even it was expected true.

soumyadeep2007 replied:
Oh I see, good catch.

haolinw commented on src/backend/access/appendonly/README.md:
```diff
@@ -286,12 +286,16 @@ With the change made by removing aoblkdir hole filling mechanism to align block
-[1] https://github.com/greenplum-db/gpdb/commit/258ec966b26929430fc5dc9f6e6fe09854644302
\ No newline at end of file
+- for indexed append-optimized tables block directory acts as source (along with visimap)
+
```
Done.

haolinw commented on src/test/regress/expected/bfv_index_optimizer.out:
```diff
@@ -485,41 +485,41 @@ set seq_page_cost=10000000;
  2 | 2 | 6 | 6
- 2 | 2 | 7 | 7
+ 1 | 1 | 5 | 5
+ 1 | 1 | 6 | 6
```
Fixed, this seems because [some additional statements in original PR](https://github.com/greenplum-db/gpdb/pull/10963/files#diff-3fb5bfbf79b2f44dcb576860d2e8c86b38c7b22c2bf3ab469a15049a22d0fe3eR341) were placed before the impact queries caused those mismatches, I removed it for useless.

soumyadeep2007 replied:
Thanks!

soumyadeep2007 commented on src/test/regress/expected/gporca.out:
```diff
@@ -13213,28 +13213,24 @@ set optimizer_enable_nljoin=off;
 -- Make sure we don't allow a regular (btree) index scan or index join for an AO table
 -- We disabled hash join, and bitmap index joins, NLJs, so this should leave ORCA no other choices
--- expect a sequential scan, not an index scan, from these two queries
+-- expect a sequential scan for GPORCA, not an index scan, from these two queries
```
Comment needs to be updated:
```
-- Make sure we don't allow a regular (btree) index scan or index join for an AO table
-- We disabled hash join, and bitmap index joins, NLJs, so this should leave ORCA no other choices
-- other than an index-only scan or sequential scan.
```

haolinw replied:
Done.

haolinw replied:
Also thank you @soumyadeep2007 for your continual review and help to get this to finish!

soumyadeep2007 reviewed (approved):
Looking good @haolinw ! Thank you for your perseverance and patience on this large PR! :heavy_check_mark: from my side.

gpopt reviewed (approved):
LGTM on Orca plan changes.

haolinw commented on src/backend/access/aocs/aocsam_handler.c:
```diff
@@ -2000,15 +2034,23 @@ aoco_relation_needs_toast_table(Relation rel)
+	 * during an index only scan.
+	 * For AO/CO tables, we never have to scan the
+	 * underlying table. This is why we set this to 1.
+	 */
```
Done.

haolinw commented on src/backend/access/appendonly/appendonlyam_handler.c:
```diff
@@ -1801,15 +1834,23 @@ appendonly_relation_needs_toast_table(Relation rel)
+	 * For AO/CO tables, we never have to scan the
+	 * underlying table. This is why we set this to 1.
+	 */
+	*allvisfrac = 1;
```
Done.
